### PR TITLE
statistics counting on gateway machines

### PR DESCRIPTION
**Pull request by janpriessnitz**

### CODE REVIEW

**Comment by janpriessnitz** [comment]
@cernvm-bot unittest

**Comment by janpriessnitz** [comment]
@jblomer So I've tried to run old gateway version (from master) with new receiver (from this PR) and the gateway crashes because it does not understand the new reply format from the receiver. Should I try to fix this, or will we require users to update to new gateway? New gateway with old receiver (the other way around) works fine by the way.

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

File: cvmfs/json_document.h [hunk 1]
```diff
@@ -13,7 +13,25 @@
 #include "util/single_copy.h"
 
 typedef struct json_value JSON;
-typedef std::vector<std::pair<const char *, const char *> > JsonStringInput;
+
+struct JsonStringInput {
+  struct JsonStringEntry {
+    std::string key;
+    std::string val;
+    bool quoted;
```
Comment (jblomer) [inline cvmfs/json_document.h]: Should be default constructed (`true`, I guess?)

**Comment by janpriessnitz** [comment]
@jlbomer The receiver<->gateway incompatibility is hitting us in the cloudtests. We need to merge this bugfix: https://github.com/cvmfs/cvmfs-gateway/pull/82 
and release a new version or do something else (I'm not sure what's the best approach).

**Comment by jblomer** [comment]
The fix is merged, and version is bumped to 1.1.1.  You should be able to build.  We then have to update the PR builder (and cloud testing job) to use the new version.  I guess for Jenkins it's enough to have a new version on ecsft.cern.ch?  Or does it need to be in the yum repos?

**Comment by janpriessnitz** [comment]
Thanks! Package on ecsft.cern.ch should be enough. I'll build it and update the cloudtesting job.

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

File: test/quickcheck/qc_catalog_merge_tool.cc [hunk 2]
```diff
@@ -181,11 +181,13 @@ RC_GTEST_FIXTURE_PROP(T_CatalogMergeTool, InOut, ()) {
 
   CatalogTestTool::History history = tester.history();
 
+  perf::Statistics *statistics = new perf::Statistics();
```
Comment (jblomer) [inline test/quickcheck/qc_catalog_merge_tool.cc]: Should be deleted somewhere

File: test/unittests/t_catalog_merge_tool.cc [hunk 3]
```diff
@@ -111,11 +111,13 @@ TEST_F(T_CatalogMergeTool, CRUD) {
 
   CatalogTestTool::History history = tester.history();
 
+  perf::Statistics *statistics = new perf::Statistics();
```
Comment (jblomer) [inline test/unittests/t_catalog_merge_tool.cc]: Stack allocation?

File: cvmfs/receiver/commit_processor.cc [hunk 4]
```diff
@@ -296,7 +298,36 @@ CommitProcessor::Result CommitProcessor::Process(
...
+        "Could not store publish statistics");
+    }
+    if (params.upload_stats_db) {
+      upload::SpoolerDefinition sd(params.spooler_configuration, shash::kAny);
+      upload::Spooler *spooler = upload::Spooler::Construct(sd);
+      if (!stats_db->UploadStatistics(spooler)) {
+        LogCvmfs(kLogReceiver, kLogSyslogErr,
+          "Could not upload statistics DB to upstream storage");
+      }
+      delete spooler;
```
[the remark this answers is not in the document]
Comment (janpriessnitz) [reply cvmfs/receiver/commit_processor.cc]: `WaitForUpload` is called inside the `UploadStatistics` function, that should be enough.

File: cvmfs/receiver/catalog_merge_tool.h [hunk 5]
```diff
@@ -37,45 +38,51 @@ class CatalogMergeTool : public CatalogDiffTool<RoCatalogMgr> {
...
+        statistics_(statistics) {}
 
   CatalogMergeTool(RoCatalogMgr* old_catalog_mgr, RoCatalogMgr* new_catalog_mgr,
                    const std::string& repo_path,
                    const PathString& lease_path,
                    const std::string& temp_dir_prefix,
                    download::DownloadManager* download_manager,
-                   manifest::Manifest* manifest)
+                   manifest::Manifest* manifest,
+                   perf::Statistics* statistics)
```
[the remark this answers is not in the document]
Comment (janpriessnitz) [reply cvmfs/receiver/catalog_merge_tool.h]: Done, but it doesn't look very nice. Do you think there is a better way?

[the remark this answers is not in the document]
Comment (jblomer) [reply cvmfs/receiver/catalog_merge_tool.h]: I think it's ok. Perhaps it's slightly nicer to initialize it with `NULL`. It should just have a well-defined value upon construction.

[the remark this answers is not in the document]
Comment (janpriessnitz) [reply cvmfs/receiver/catalog_merge_tool.h]: Ok, thanks, `NULL` looks better.

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

**Comment by janpriessnitz** [comment]
@jblomer I guess this one is good to go. I've addressed the review comments. Let me know if there's something else to be done.

**Comment by janpriessnitz** [comment]
@cernvm-bot cloudtest

**Pull request merged**